File: 프로그래머스/3/42861. 섬 연결하기/섬 연결하기.py
import heapq
def solution(n, costs):
    answer = 0
    reachable = [[] for _ in range(n)]
    
    for edge in costs:
        reachable[edge[0]].append((edge[1], edge[2]))
        reachable[edge[1]].append((edge[0], edge[2]))
        
    connected = set()
    pq = [(0, 0)]
    
    while True:
        if len(connected) == n:
            break
            
        cost, node = heapq.heappop(pq)
        if node in connected:
            continue

        connected.add(node)
        answer += cost

        for next_node, next_cost in reachable[node]:
            if next_node not in connected:
                heapq.heappush(pq, (next_cost, next_node))
                
    return answer

File: 프로그래머스/3/42861. 섬 연결하기/섬 연결하기.java
import java.util.*;
class Solution {
    public int solution(int n, int[][] costs) {
        int answer = 0;

    PriorityQueue<int[]> pq = new PriorityQueue<>((e1, e2) -> e1[0] - e2[0]);
    List<Integer>[] canGoFrom = new ArrayList[n];
    boolean[] visited = new boolean[n];

    for (int i = 0; i < n; i++) {
      visited[i] = false;
      canGoFrom[i] = new ArrayList<>();
    }

    for (int i = 0; i < costs.length; i++) {
      canGoFrom[costs[i][0]].add(i);
      canGoFrom[costs[i][1]].add(i);
    }


    visited[0] = true;
    for (int i = 0; i < canGoFrom[0].size(); i++) {
      int edgeIdx = canGoFrom[0].get(i);
      int toNode = Math.max(costs[edgeIdx][0], costs[edgeIdx][1]);
      pq.add(new int[] {costs[edgeIdx][2], toNode});
    }

    while (!pq.isEmpty()) {
      int cost = pq.peek()[0];
      int nextNode = pq.peek()[1];
      pq.poll();

      if (!visited[nextNode]) {
        visited[nextNode] = true;
        answer += cost;
        for (int i = 0; i < canGoFrom[nextNode].size(); i++) {
          int edgeIdx = canGoFrom[nextNode].get(i);
          int toNode = costs[edgeIdx][0] + costs[edgeIdx][1] - nextNode;
          pq.add(new int[] {costs[edgeIdx][2], toNode});
        }
      }
    }

    return answer;
    }
}